core/App/ffi_operation.cpp

```diff
@@ -390,6 +390,298 @@ extern "C"
+    uint32_t ffi_importKeyMaterial(JsonObj payloadJson, char *respJson)
+    {
+        RetJsonObj retJsonObj;
+
+        JsonObj tokenJsonObj;
+
+        ehsm_keyblob_t *cmk = NULL;
+        ehsm_keyblob_t *sessionkeyblob = NULL;
+        ehsm_status_t ret = EH_OK;
+        ehsm_data_t *key_material = NULL;
+        ehsm_data_t *hmac = NULL;
+        ehsm_data_t *importToken = NULL;
+        std::string hmac_str;
+
+        JSON2STRUCT(payloadJson, cmk);
+        JSON2STRUCT(payloadJson, sessionkeyblob);
+
+        std::string token_str = base64_decode(payloadJson.readData_string("importToken"));
+        tokenJsonObj.parse(token_str);
+
+        std::string signature = tokenJsonObj.readData_string("hmac");
+        uint32_t time_token = atoi(tokenJsonObj.readData_cstr("timestamp"));
+        uint32_t timestamp = atoi(payloadJson.readData_cstr("timestamp"));
+
+        Json::Value token_json = tokenJsonObj.getJson();
+        token_json.removeMember("hmac");
+        tokenJsonObj.setJson(token_json);
+
+        importToken = (ehsm_data_t *)malloc(APPEND_SIZE_TO_DATA_T(tokenJsonObj.toString().size() - 1));
+        importToken->datalen = APPEND_SIZE_TO_DATA_T(tokenJsonObj.toString().size() - 1);
+
+        memcpy_s(importToken->data, importToken->datalen, tokenJsonObj.toString().c_str(), importToken->datalen);
+
+        string key_material_str_base64 = payloadJson.readData_string("key_material");
+
+        string key_material_str = base64_decode(key_material_str_base64);
+
+        key_material = (ehsm_data_t *)malloc(APPEND_SIZE_TO_DATA_T(key_material_str.length()));
+        key_material->datalen = key_material_str.length();
+        memcpy_s(key_material->data, key_material->datalen, key_material_str.c_str(), key_material->datalen);
+
+        ehsm_padding_mode_t padding_mode = (ehsm_padding_mode_t)payloadJson.readData_uint32("padding_mode");
+
+        // generate hmac space
+        hmac = (ehsm_data_t *)malloc(APPEND_SIZE_TO_DATA_T(EH_HMAC_SHA256_SIZE));
+        if (hmac == NULL)
+        {
+            retJsonObj.setCode(retJsonObj.CODE_FAILED);
+            retJsonObj.setMessage("Server exception.");
+            goto out;
+        }
+        hmac->datalen = EH_HMAC_SHA256_SIZE;
+
+        if (sessionkeyblob == NULL)
+        {
+            retJsonObj.setCode(retJsonObj.CODE_FAILED);
+            retJsonObj.setMessage("SessionKeyBlob is empty.");
+            goto out;
+        }
+
+        ret = GenerateTokenHmac(sessionkeyblob, importToken, hmac);
+        if (ret != EH_OK)
+        {
+            retJsonObj.setCode(retJsonObj.CODE_FAILED);
+            retJsonObj.setMessage("Server exception.");
+            goto out;
+        }
+
+        hmac_str = base64_encode(hmac->data, hmac->datalen);
+        if (hmac_str != signature)
+        {
+            retJsonObj.setCode(retJsonObj.CODE_FAILED);
+            retJsonObj.setMessage("Server exception.");
+            goto out;
+        }
+        if (timestamp > time_token + 86400000)
```

Comment:
86400000 means?

Reply:
24 hours, is there a better way to express it, as a macro?